kore/src/Kore/Attribute/Pattern/NonSimplifiable.hs

```diff
@@ -0,0 +1,191 @@
+{- |
+Copyright   : (c) Runtime Verification, 2019
+License     : NCSA
+
+ -}
+
+module Kore.Attribute.Pattern.NonSimplifiable
+    ( NonSimplifiable (..)
+    ) where
+
+import Control.DeepSeq
+import Data.Hashable
+import qualified Generics.SOP as SOP
+import qualified GHC.Generics as GHC
+
+import Kore.Attribute.Synthetic
+import Kore.Debug
+import Kore.Domain.Builtin
+import Kore.Internal.Alias
+    ( Alias
+    )
+import Kore.Internal.Symbol
+    ( Symbol
+    )
+import qualified Kore.Internal.Symbol as Symbol
+import Kore.Syntax
+import Kore.Syntax.Application
+    ( Application (..)
+    )
+import Kore.Variables.UnifiedVariable
+    ( UnifiedVariable
+    )
+
+{- | A pattern is `NonSimplifiable` if:
+    1. it's a `BuiltinBool`, `BuiltinInt` or a `BuiltinString`
```

Comment:
@ana-pantilie You asked me about this before. I'm sorry I only realized this now:

@ana-pantilie @ttuegel Should we also include domain values of unrecognized types (i.e. not int, bool, string) here? I think (but I may be wrong) that imp uses such domain values for IDs, and I think it uses them as storage keys.

Also, strictly speaking, we could probably include sets, lists and maps when all their elements are non-simplifiable, but it's probably not worth the trouble.

Reply:
@virgil-serbanuta Yes, we use domain values as keys as well. Sorry, I marked this pull request as ready before testing the integration tests with the other PR as well. Why do you think we should only include domain values of unrecognized types?

Reply:
We could think about that, but the only kind of recognizable domain values that I saw had a string child, and we usually check that it's a valid string. And we simplify those to BuiltinBool/Int/String.

I did try to find hypothetical examples of bool/int/string domain values that do not have strings as children, and which are unsimplifiable at the same time, and I'm not sure that makes sense (i.e. I would expect that any non-string child would be simplifiable to a string), but I'm not fully sure yet.

Reply:
> I think (but I may be wrong) that imp uses such domain values for IDs, and I think it uses them as storage keys.

This is correct.

> Also, strictly speaking, we could probably include sets, lists and maps when all their elements are non-simplifiable, but it's probably not worth the trouble.

Agreed.

Reply:
> Why do you think we should only include domain values of unrecognized types and not all domain values with a non-simplifiable child?

During the daily meeting we checked and found that for all the recognized types of domain value, the pattern verifier checks that they have a literal string argument and decodes them to their builtin representation; it is an error otherwise. So, it should be safe to assume that any `\dv` remaining during execution is in an unrecognized sort.